pyrsistent/_pmap.py

```diff
@@ -125,7 +125,27 @@ def __len__(self):
     def __repr__(self):
         return 'pmap({0})'.format(str(dict(self)))
 
-    __eq__ = Mapping.__eq__
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        if len(self) != len(other):
+            return False
+        if isinstance(other, PMap):
+            try:
+                if hash(self) != hash(other):
+                    return False
+            except TypeError:
+                # self or other contains non-hashable values
+                pass
+            if self._buckets == other._buckets:
+                return True
+            return dict(self.iteritems()) == dict(other.iteritems())
```

Comment:
> In the case you describe, I think we'd have to traverse two, not three times, because a different number of buckets would mean O(1) for checking list equality, because the lengths of the lists differ.

Sorry, thought it was a list of buckets, but it's actually a vector, so this may not be true.